PlayList.java

```diff
@@ -30,101 +32,201 @@ public Track getTrack(int index) {
-     *  is full, does nothing and returns false. Otherwise, inserts the track and
-     *  returns true. */
+    /**
+     * Inserts the given track in index i of this list. For example, if the list is
+     * (t5, t3, t1), then just after add(1,t4) the list becomes (t5, t4, t3, t1).
+     * If the list is the empty list (), then just after add(0,t3) it becomes (t3).
+     * If i is negative or greater than the size of this list, or if the list
+     * is full, does nothing and returns false. Otherwise, inserts the track and
+     * returns true.
+     */
     public boolean add(int i, Track track) {
         //// replace the following statement with your code
+        if (i <= this.size && i >= 0 && this.size <= this.maxSize) {
+            if (i == this.size) {
+                this.add(track);
+            } else {
+                for (int j = this.size; j > i; j--) {
+                    this.tracks[j] = this.tracks[j - 1];
+                }
+                this.tracks[i] = track;
+                this.size++;
+            }
+            return true;
+        }
+
         return false;
     }
-     
-    /** Removes the track in the given index from this list.
-     *  If the list is empty, or the given index is negative or too big for this list, 
-     *  does nothing and returns -1. */
+
+    /**
+     * Removes the track in the given index from this list.
+     * If the list is empty, or the given index is negative or too big for this
+     * list,
+     * does nothing and returns -1.
+     */
     public void remove(int i) {
         //// replace this comment with your code
+        if (this.size == 0 || i < 0 || i >= this.size) {
+            return;
+        }
+        if (i == this.size - 1) {
+            this.removeLast();
+        } else {
+            for (int j = i; j < this.size - 1; j++) {
+                this.tracks[j] = this.tracks[j + 1];
+            }
+            this.tracks[this.size - 1] = null;
+            this.size--;
+        }
     }
 
-    /** Removes the first track that has the given title from this list.
-     *  If such a track is not found, or the list is empty, or the given index
-     *  is negative or too big for this list, does nothing. */
+    /**
+     * Removes the first track that has the given title from this list.
+     * If such a track is not found, or the list is empty, or the given index
+     * is negative or too big for this list, does nothing.
+     */
     public void remove(String title) {
         //// replace this comment with your code
+        this.remove(indexOf(title));
     }
 
-    /** Removes the first track from this list. If the list is empty, does nothing. */
+    /**
+     * Removes the first track from this list. If the list is empty, does nothing.
+     */
     public void removeFirst() {
         //// replace this comment with your code
+        if (this.size == 0) {
+            return;
+        } else {
+            this.remove(0);
+        }
```

Comment:
remove(index) do the same validations, you can just call it :)
```suggestion
            this.remove(0);
```